KomaMRICore/src/simulation/GPUFunctions.jl

```diff
@@ -42,7 +42,8 @@ _isleaf(x) = _isbitsarray(x) || isleaf(x)
 struct KomaCUDAAdaptor end
 adapt_storage(to::KomaCUDAAdaptor, x) = CUDA.cu(x)
 # SimpleMotion
-adapt_storage(to::KomaCUDAAdaptor, x::SimpleMotionType) = adapt(Float32, x) # SimpleMotionType is not stored in GPU
+adapt_storage(to::KomaCUDAAdaptor, x::SimpleMotion) = SimpleMotion(paramtype.(Float32, x.types)) 
```

Comment:
What sould I do, then?
Passing SimpleMotionTypes to GPU does not make sense, since only scalar values are stored in this structures.

Reply:
Why can't you do something like this `adapt_storage(T::Type{<:Real}, xs::SimpleMotion) = SimpleMotion(paramtype.(T, xs.types))`

Reply:
That is done in line 107, for Precission conversion (`f32`, `f64`). The problem is that when `gpu` function is used, it automatically converts the inner type of structures to Float32 (I don't know if this is the expected behaviour). As `SimpleMotion` and `SimpleMotionType` cannot be passed to gpu, I have forced the conversion to float32 so that there is no type mismatch.

Reply:
I know that is for precision, but you should not force the type to Float32. The GPU can use Flaot64. If `gpu` converts to Float32 it means something is wrong with the adaptors, as I did not experience this.

Reply:
It might be a bug, because in master branch, it also occurs:
`phantom = brain_phantom2D()`
Phantom{Float64}
  name: String "brain2D_axial"
  x: Array{Float64}((6506,)) [-0.084, -0.084, -0.084, -0.084, -0.084, -0.084, -0.084, -0.084, -0.084, -0.084  …  0.084, 0.084, 0.084, 0.084, 0.086, 0.086, 0.086, 0.086, 0.086, 0.086]
  y: Array{Float64}((6506,)) [-0.03, -0.028, -0.026, -0.024, -0.022, -0.02, -0.018, -0.016, -0.014, -0.012  …  0.006, 0.008, 0.01, 0.012, -0.008, -0.006, -0.004, -0.002, 0.0, 0.002]
  z: Array{Float64}((6506,)) [-0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0  …  0.0, 0.0, 0.0, 0.0, -0.0, -0.0, -0.0, -0.0, 0.0, 0.0]
  ρ: Array{Float64}((6506,)) [0.7, 0.7, 0.7, 0.7, 0.7, 0.7, 0.7, 0.7, 0.7, 0.7  …  0.7, 0.7, 0.7, 0.7, 0.7, 0.7, 0.7, 0.7, 0.7, 0.7]
  T1: Array{Float64}((6506,)) [0.5690000000000001, 0.5690000000000001, 0.5690000000000001, 0.5690000000000001, 0.5690000000000001, 0.5690000000000001, 0.5690000000000001, 0.5690000000000001, 0.5690000000000001, 0.5690000000000001  …  0.5690000000000001, 0.5690000000000001, 0.5690000000000001, 0.5690000000000001, 0.5690000000000001, 0.5690000000000001, 0.5690000000000001, 0.5690000000000001, 0.5690000000000001, 0.5690000000000001]
  T2: Array{Float64}((6506,)) [0.329, 0.329, 0.329, 0.329, 0.329, 0.329, 0.329, 0.329, 0.329, 0.329  …  0.329, 0.329, 0.329, 0.329, 0.329, 0.329, 0.329, 0.329, 0.329, 0.329]
  T2s: Array{Float64}((6506,)) [0.058, 0.058, 0.058, 0.058, 0.058, 0.058, 0.058, 0.058, 0.058, 0.058  …  0.058, 0.058, 0.058, 0.058, 0.058, 0.058, 0.058, 0.058, 0.058, 0.058]
  Δw: Array{Float64}((6506,)) [-0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0  …  -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0]
  Dλ1: Array{Float64}((6506,)) [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
  Dλ2: Array{Float64}((6506,)) [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
  Dθ: Array{Float64}((6506,)) [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
  ux: #115 (function of type KomaMRIBase.var"#115#130")
  uy: #116 (function of type KomaMRIBase.var"#116#131")
  uz: #117 (function of type KomaMRIBase.var"#117#132")

`phantom |> gpu`
Phantom{Float32}
  name: String "brain2D_axial"
  x: CUDA.CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}
  y: CUDA.CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}
  z: CUDA.CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}
  ρ: CUDA.CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}
  T1: CUDA.CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}
  T2: CUDA.CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}
  T2s: CUDA.CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}
  Δw: CUDA.CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}
  Dλ1: CUDA.CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}
  Dλ2: CUDA.CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}
  Dθ: CUDA.CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}
  ux: #115 (function of type KomaMRIBase.var"#115#130")
  uy: #116 (function of type KomaMRIBase.var"#116#131")
  uz: #117 (function of type KomaMRIBase.var"#117#132")

Reply:
It seems that the problem comes from `CUDA.cu()` function:

`using CUDA`
`a = [0.1, 0.2, 0.3]`
3-element Vector{Float64}:
 0.1
 0.2
 0.3

`gpu_a = cu(a)`
3-element CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}:
 0.1
 0.2
 0.3

If we use `CuArray()` instead:
`gpu_a = CuArray(a)`
3-element CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}:
 0.1
 0.2
 0.3